### PR TITLE
refactor(api-request-builder): pass project key as main option

### LINKED ISSUE #43: [api-request-builder] pass projectKey as main option
Instead of passing it every time as option of `.build()`, pass it once to `createRequestBuilder`.

> It's a BREAKING CHANGE

### PR DESCRIPTION
**Pull request by wizzy25**
#### Summary

affects: @commercetools/api-request-builder, @commercetools/csv-parser-price

### BREAKING CHANGE!!!
pass the projectKey as a required option to the ```createRequestBuilder``` method

resolves #43

### CODE REVIEW

**Comment by wizzy25** [comment]
All builds for this repo are failing, including master @emmenko @hisabimbola

**Comment by emmenko** [comment]
@wizzy25 looks like the integration tests are failing?

**Comment by wizzy25** [comment]
@emmenko They're also failing on master... Reverting commits to see if/where something broke

File: packages/api-request-builder/src/build-query-string.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 /* @flow */
 import type {
   ServiceBuilderDefaultParams,
-} from 'types/sdk'
+} from '../../../types/sdk'
```
Comment (emmenko) [inline packages/api-request-builder/src/build-query-string.js]: Why this change?

File: packages/api-request-builder/src/create-service.js [hunk 2]
```diff
@@ -30,6 +31,7 @@ const requiredDefinitionProps = [
 
 export default function createService (
   definition: ServiceBuilderDefinition,
+  project: BuildOptions = {},
```
Comment (emmenko) [inline packages/api-request-builder/src/create-service.js]: This should be called `options`

File: packages/api-request-builder/src/create-service.js [hunk 3]
```diff
@@ -42,7 +44,12 @@ export default function createService (
   if (!Array.isArray(definition.features) || !definition.features.length)
     throw new Error('Definition requires `features` to be a non empty array.')
 
+  if (!project.projectKey)
+    throw new Error('No project defined. Please enter a project key')
```
Comment (emmenko) [inline packages/api-request-builder/src/create-service.js]: You are validating twice (here and in `create-request-builder.js`). Just pick one ;)

File: packages/api-request-builder/test/create-service.spec.js [hunk 4]
```diff
@@ -66,7 +73,7 @@ describe('createService', () => {
     )
   })
 
-  it('should throw if definition prop (features) is a non empty array', () => {
+  it('should throw if definition prop (features) is an empty array', () => {
```
Comment (emmenko) [inline packages/api-request-builder/test/create-service.spec.js]: Why this change?

File: types/sdk.js [hunk 5]
```diff
@@ -124,6 +124,9 @@ export type ServiceBuilderDefinition = {
 export type ApiRequestBuilder = {
   [key: string]: ServiceBuilder;
 }
+export type BuildOptions = {
```
Comment (emmenko) [inline types/sdk.js]: We might find a bit more specific name

File: packages/api-request-builder/src/build-query-string.js (hunk 1 above)
Comment (wizzy25) [inline packages/api-request-builder/src/build-query-string.js]: ```types/sdk``` gives this lint error ```unable to resolve path to module import/no-unresolved```, and manually tracing it resolves to the types folder in the root ```../../../types/sdk``` (which gives no errors)

File: packages/api-request-builder/src/create-service.js (hunk 3 above)
Comment (wizzy25) [inline packages/api-request-builder/src/create-service.js]: Thought so too... But is there a scenario where ```createService``` can be called on it's own? (not in ```create-request-builder```)? Or I'll just do the validation here so it's throws no matter where it's called ;-)

File: packages/api-request-builder/test/create-service.spec.js (hunk 4 above)
Comment (wizzy25) [inline packages/api-request-builder/test/create-service.spec.js]: The old one isn't correct ;)

File: types/sdk.js (hunk 5 above)
Comment (wizzy25) [inline types/sdk.js]: Didn't want to really mess with the variable names, and since [you feel the variable should be called ```options```](https://github.com/commercetools/nodejs/pull/116#discussion_r108428216), what better could there be?

File: packages/api-request-builder/src/build-query-string.js (hunk 1 above)
Comment (emmenko) [inline packages/api-request-builder/src/build-query-string.js]: This was working fine, there is an alias resolver for that (otherwise linting on master will fail). Are you sure you have the correct dependencies installed?

File: packages/api-request-builder/src/create-service.js (hunk 3 above)
Comment (emmenko) [inline packages/api-request-builder/src/create-service.js]: Nope, only `create-request-builder` is exposed

Comment (emmenko) [inline packages/api-request-builder/src/create-service.js]: And yes, you can just do it here in `create-service.js` ;)

File: packages/api-request-builder/test/create-service.spec.js (hunk 4 above)
Comment (emmenko) [inline packages/api-request-builder/test/create-service.spec.js]: 🤔 why? `features` needs to be an array, so if it's **a non** empty array it throws

File: types/sdk.js (hunk 5 above)
Comment (emmenko) [inline types/sdk.js]: Something like `ApiRequestBuilderServiceOptions`

File: packages/api-request-builder/src/build-query-string.js (hunk 1 above)
Comment (wizzy25) [inline packages/api-request-builder/src/build-query-string.js]: Still having the errors straight from a fresh git clone on master.
More like a warning though as everything still "seems" to work fine as it is...

Comment (emmenko) [inline packages/api-request-builder/src/build-query-string.js]: Can you post here the output of `npm run lint`?

File: packages/api-request-builder/test/create-service.spec.js (hunk 4 above)
Comment (wizzy25) [inline packages/api-request-builder/test/create-service.spec.js]: In that case the error message is wrong: 
>Definition requires `features` to be a non empty array

Should instead be: 
>Definition requires `features` to be an empty array

...and [here](https://github.com/commercetools/nodejs/blob/43-projectKey-main-option/packages/api-request-builder/test/create-service.spec.js#L83), you still expect it to throw when features is **empty** 😏

Comment (emmenko) [inline packages/api-request-builder/test/create-service.spec.js]: Ok, maybe we should split the test into 2 cases then ;)

File: packages/api-request-builder/src/build-query-string.js (hunk 1 above)
Comment (hisabimbola) [inline packages/api-request-builder/src/build-query-string.js]: @wizzy25 it is a sublime specific error. Info how to fix it [here](https://github.com/benmosher/eslint-plugin-import/blob/master/README.md#sublimelinter-eslint)

Comment (emmenko) [inline packages/api-request-builder/src/build-query-string.js]: Please revert the changes then ;)

File: packages/csv-parser-price/test/main.spec.js [hunk 6]
```diff
@@ -16,10 +16,13 @@ const logger = {
   info: () => {},
   verbose: () => {},
 }
+const apiConfig = {
```
Comment (hisabimbola) [inline packages/csv-parser-price/test/main.spec.js]: I would have prefered if this refactor is done in another PR...

Comment (wizzy25) [inline packages/csv-parser-price/test/main.spec.js]: yeah but then, the tests would fail

File: packages/csv-parser-price/src/main.js [hunk 7]
```diff
@@ -183,9 +183,11 @@ export default class CsvParserPrice {
 // Easiest way to wrap the getCustomTypeDefinition in the memoize method
 CsvParserPrice.prototype.getCustomTypeDefinition = memoize(
   function _getCustomTypeDefinition (customTypeKey) {
-    const getTypeByKeyUri = createRequestBuilder().types
+    const getTypeByKeyUri = createRequestBuilder({
```
Comment (hisabimbola) [inline packages/csv-parser-price/src/main.js]: Why are we changing this?

Comment (wizzy25) [inline packages/csv-parser-price/src/main.js]: Because we pass the projectkey to `.build()` and now we will be  passing it to `createRequestBuilder()`. 
That's what this PR is about 🙄

Comment (wizzy25) [inline packages/csv-parser-price/src/main.js]: That's why it's a breaking change

Comment (hisabimbola) [inline packages/csv-parser-price/src/main.js]: Yeah, but that did not fully explain why a code in another package affects this...

Comment (wizzy25) [inline packages/csv-parser-price/src/main.js]: This module [uses](https://github.com/commercetools/nodejs/blob/master/packages/csv-parser-price/src/main.js#L13) the `createRequestBuilder()` method, which is part of the `api-request-builder`, which is this PR

Comment (hisabimbola) [inline packages/csv-parser-price/src/main.js]: Yeah, the fact that it uses it doesn't still validate the change

Do we change the module everytime another package is updated?

`csv-parser-price` uses `lodash` for example, sure we don't update our module everytime that gets updated

So you should see where I am driving to with this...

Comment (wizzy25) [inline packages/csv-parser-price/src/main.js]: So what do you suggest? Because I don't know how we can pass and not pass the projectkey to the `.build()` method at the same time

Comment (hisabimbola) [inline packages/csv-parser-price/src/main.js]: A more appropriate reason can be found [here](https://github.com/lerna/lerna#how-bootstrap-works)

Lerna works by symlinking internal modules when you do 'npm install', so when you make a change to a module, all other modules get's the new updated change, and this is the desired behaviour

**Comment by hisabimbola** [comment]
I think this is good to go @emmenko 

Can you take a final look and let's ship this 🚀

**Comment by wizzy25** [comment]
@emmenko updated docs. Please review

File: docs/sdk/api/apiRequestBuilder.md [hunk 8]
```diff
@@ -14,27 +14,31 @@ npm install --save @commercetools/api-request-builder
 <script>// global: CommercetoolsApiRequestBuilder</script>
 ```
 
-## `createRequestBuilder(customServices)`
+## `createRequestBuilder(projectKey, customServices)`
```
Comment (emmenko) [inline docs/sdk/api/apiRequestBuilder.md]: Sorry, now that I see this I think it's better to have named arguments

```js
// before
createRequestBuilder(projectKey, customServices)

// after
createRequestBuilder(options)
// options -> { projectKey, customServices }
```

/cc @hisabimbola ?

Comment (wizzy25) [inline docs/sdk/api/apiRequestBuilder.md]: Do you mean to refactor `createRequestBuilder()` to take in one object as argument containing projectKey and customServices? 😶 🙄

Comment (hisabimbola) [inline docs/sdk/api/apiRequestBuilder.md]: @wizzy25 and I actually discussed this format, but can't remember why we didn't go with it. We can refactor it. Looks better-managed params

Comment (emmenko) [inline docs/sdk/api/apiRequestBuilder.md]: Cool, thanks and sorry for dragging this longer 😁

**Comment by wizzy25** [comment]
@emmenko Hope this would be last time 😄 
kindly review

File: docs/sdk/api/apiRequestBuilder.md [hunk 9]
```diff
@@ -14,38 +14,47 @@ npm install --save @commercetools/api-request-builder
...
 
 #### Arguments
 
-1. `customServices` *(Object)*: A map of custom services that are not provided by default. This might be useful to build a request for a different API with similar query parameters.
+The `options` argument must be an object with a `projectKey` property, and optionally a `customServices` property
+```js
+    options -> { projectKey, customServices }
+```
+
+1. `projectKey` *(String)*: A required staring specifying the project key to use for the request, with the format. Even though this is required, the project key can be omitted from the URI by passing `{withProjectKey: false}` to `.build()`
```
Comment (emmenko) [inline docs/sdk/api/apiRequestBuilder.md]: `staring`?

Comment (wizzy25) [inline docs/sdk/api/apiRequestBuilder.md]: oops... typo. 🙈 
But is that the only thing? So I could fix it all in one commit

**Comment by wizzy25** [comment]
LGTM!!! 
@hisabimbola @emmenko
